tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/queryfixer/fixer/NearbyTokenFixer.java

```diff
@@ -0,0 +1,136 @@
+import com.google.cloud.bigquery.utils.queryfixer.entity.IToken;
+import com.google.cloud.bigquery.utils.queryfixer.entity.Position;
+import com.google.cloud.bigquery.utils.queryfixer.errors.ExpectKeywordButGotOthersError;
+import com.google.cloud.bigquery.utils.queryfixer.tokenizer.QueryTokenProcessor;
+import com.google.cloud.bigquery.utils.queryfixer.util.StringUtil;
+import com.google.common.collect.ImmutableList;
+import org.apache.commons.lang3.tuple.Pair;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * A class to fix general syntax errors. Usually, a general syntax error looks like {@link
+ * ExpectKeywordButGotOthersError} but expects "end of input". It basically means there may be
+ * multiple expected tokens but the error message cannot recognize them, so the fixer will try to
+ * see if the tokens near the error position are similar to any keywords. If yes, then replace the
+ * token to the similar keyword anc check if the error is eliminated from the query.
+ *
+ * <p>Here is an example with an input query:
+ *
+ * <pre>
+ *     SELECT status FORM `bigquery-public-data.austin_311.311_request` LIMIT 10
+ * </pre>
+ *
+ * It causes an error "Syntax error: Expected end of input but got identifier `...` at [1:20]". The
+ * fixer will look around this position and find FORM looks like a keyword FROM and convert it to
+ * FROM, leading to a new query
+ *
+ * <pre>
+ *     SELECT status FROM `bigquery-public-data.austin_311.311_request` LIMIT 10
+ * </pre>
+ */
+public class NearbyTokenFixer implements IFixer {
+
+  private final String query;
+  private final ExpectKeywordButGotOthersError err;
+  private final QueryTokenProcessor queryTokenProcessor;
+  private final List<String> keywords;
+
+  // TODO: it could be configured by users in future.
+  private static final double SIMILARITY_THRESHOLD = 0.5;
+
+  public NearbyTokenFixer(
+      String query, ExpectKeywordButGotOthersError err, QueryTokenProcessor queryTokenProcessor) {
+    this.query = query;
+    this.err = err;
+    this.queryTokenProcessor = queryTokenProcessor;
+    this.keywords = getAllKeywords();
+  }
+
+  @Override
+  public FixResult fix() {
+
+    Position errorPosition = err.getErrorPosition();
+    Pair<IToken, IToken> tokens =
+        queryTokenProcessor.getNearbyTokens(
+            query, errorPosition.getRow(), errorPosition.getColumn());
+
+    StringUtil.SimilarStrings rightTokenSimilarStrings = findSimilarKeywords(tokens.getRight());
+    StringUtil.SimilarStrings leftTokenSimilarStrings = findSimilarKeywords(tokens.getLeft());
+
+    // The overall logic below is that if left token has a more similar (measured by
+    // edit distance) keyword, then only choose the fixes on left tokens. If right token
+    // has more similar keywords, choose the fixes on right tokens. If both of them have
+    // the similar keywords with the same edit distance, then choose both of them.
```

Comment:
Do we have a use case where this would be confident? where left edit distance = right distance and use both of them?

Reply:
I think it is hard to consider them as confident, because we are guessing the correct keywords rather than inferring one from the error message.